docs/WritingTests.asciidoc

```diff
@@ -22,52 +22,208 @@ https://github.com/os-autoinst/openQA[official repository].
-take a look to the published documentation at http://open.qa/api/testapi/. This test API
-is sometimes also referred to as an openQA DSL, because in some contexts it can look like
-a domain specific language.
+{testapi} provides the API for the test modules using the os-autoinst backend.
+Take a look at the http://open.qa/api/testapi[test API documentation] for further
+information. Note that this test API is sometimes also referred to as an
+openQA DSL, because in some contexts it can look like a domain specific
+language.
+
+=== Notes on the Python API
+[id="notes-python-api"]
+
+The Python integration that OpenQA offers through `Inline::Python` also allows
+the test modules to import other Perl modules with the usage of the `perl`
+virtual package provided by `Inline::Python`.
+
+Because of the way `Inline::Python` binds Perl functions to Python it is not
+possible to use keywords arguments from Python to Perl functions. They must be
+passed as positional arguments, for example `"key", "value"`.
+
+See the following snippet of Perl code
+
+[source,perl]
+-------------------------------------------------------------------
+use x11utils;
+
+# [...] omitted for brevity
+
+sub run() {
+    # [...] omitted for brevity
+
+    # Start vncviewer - notice the named arguments
+    x11_start_program('vncviewer :0',
+        target_match => 'virtman-gnome_virt-install',
+        match_timeout => 100
+    );
+    # [...] omitted for brevity
+}
+
+1;
+-------------------------------------------------------------------
+
+versus the equivalent python code:
+
+[source,python]
+-------------------------------------------------------------------
+from testapi import *
+
+# [...] omitted for brevity
+
+def run():
+    perl.require('x11utils')
+
+    # Start vncviewer - notice the named arguments passed as positional arguments
+    # Formatted in pairs for better visibility.
+
+    perl.x11utils.x11_start_program('vncviewer :0',
+        'target_match', 'virtman-gnome_virt-install',
+        'match_timeout', 100
+    )
+    # [...] omitted for brevity
+-------------------------------------------------------------------
+
+== How to write test modules
+
+=== Test module interface
+[caption="Example: "]
+.Note: symbols marked with `*1` are optional. They are no-ops when not present.
+
+All openQA tests need to comply with the following interface (loosely defined in
+type-hinted python terms for the sake of explicitness):
+
+[source,python]
+-------------------------------------------------------------------
+def run(): -> None
+def test_flags(): -> dict # *1
+def post_fail_hook(): -> None # *1
+def pre_run_hook(): -> None # *1
+def post_run_hook(): -> None # *1
+-------------------------------------------------------------------
+
+==== `run`
+Defines the actual steps to be performed during the module execution.
+
+An example usage:
+
+[source,perl]
+-------------------------------------------------------------------
+sub run() {
```

Comment:
I don't understand your change here. First you explain the interface that python test modules must implement and then the perl run method is shown as example? Also the python test modules do *not* need to implement those functions, only `run` is required. For the rest the baseclass already provide implementations

Reply:
The desire is to layout the complete interface of a test module in a language/dialect/markup that allows to indicate: signatures & return types.

The interface is not "defined" in python per se, but rather explained in type-hinted python terms (so we can get signatures & return types) on what's actually required for the test.

I could layout the base class and what can be overwritten there. But there's no "class" in the actual test module, or is it? (I still get lost in perl, please correct me if my assumption is wrong)

Reply:
> But there's no "class" in the actual test module, or is it?

There is, https://github.com/os-autoinst/os-autoinst/blob/master/basetest.pm

Reply:
> > But there's no "class" in the actual test module, or is it?
> 
> There is, https://github.com/os-autoinst/os-autoinst/blob/master/basetest.pm

Basetest is a poor example, as it is tightly coupled with code that should be on the backend baseclass, autotest or testapi (i.e, anything that calls query_isotovideo): https://github.com/os-autoinst/os-autoinst/blob/9a99cd728b32edc918b2d170e947eecd6638656d/basetest.pm#L624

Reply:
After a deep dive into understanding "classes" in perl. Any sub defined in [basetest.pm][basetest.pm] could potentially be overwritten by the test module and it's allowed.

[basetest.pm][basetest.pm] is an Abstract-ish Base "Class", a concrete "class" and the test context at the same time :sweat_smile:.

The thing is now that the way these classes are used is *very* perl, they can't be called classes in python because Python classes are a completely different beast.

Right now I can't think of another way to markup the signature of what's expected for a developer to provide in a test module that what's in the docs in 995dbad4f, but I'm open to anything neutral that indicates:

* Function/Subroutine Signature
* Return Types
* What symbols are optional

[basetest.pm]: https://github.com/os-autoinst/os-autoinst/blob/master/basetest.pm

Reply:
Well, the issue I see right now is that your additions state that a test module would need to implement an interface. As a software designer I understand it and it's correct but the target audience are testers and QA engineers with little to no programming experience. Hence we should be careful to use mostly plain English even under the risk of not completely technically correct statements. Formerly the text stated something along the lines of "To write a test module just put a run-function and you are good". That's what I would like to keep.

Reply:
>  As a software designer I understand it and it's correct but the target audience are testers and QA engineers with little to no programming experience.

This assumption is incorrect (and a bit condescending), I would be ok if `interface` links to a definition on I.E Wikipedia, while we build a glossary of terms that we can refer to later. Also something missing is to say that the only required method is the run method. I'd buy (barely) Perl, but not a general statement such as that

The target audience is QA Engineers, who regardless of their programming experience have decided to look into test automation using openQA.

Reply:
> Hence we should be careful to use mostly plain English even under the risk of not completely technically correct statements.

I think in this case we can keep both, check 8cd76802f.

Brought back the former writing focusing on `run` and elaborated below so it's clear that only a single function is needed.